Update shared resource to use atomic references instead of locks.

jsuereth opened the pull request:
Lockless would be ideal for these, but there's a lot of caveats in the implementation that need to be thought through, specifically "close after exception".

Review by: @Atry

Atry commented on src/main/scala/resource/SharedResource.scala:
```diff
@@ -0,0 +1,59 @@
+		if (myCount == 1) {
+			// We must construct the reference
+			// We may want to CAS here and fail if we're inconsistent.
+			val r = opener
```
What if `opener` throws an exception?

jsuereth replied:
Good point.  That will literally cause all threads waiting for the resource to be blocked.  This confirms my suspicion we need some kind of an error channel to signal to all threads that the resource has failed and they need to bail.   Let me think about how to handle this appropriately....

Atry replied:
When one thread failed to open a resource, what is the expected behavior for other tries from other threads?
One possible solution is `synchronized` the entire `open` method, including the `addAndGet`. So that it reset count to zero if fails, and let other threads retry `openr`.

Atry replied:
If the entire `open` method is `synchronized`, no error channel is required.

Atry replied:
Note that my original implementation increase the `count` after `openr`. Hence I did not have to catch the exception and reset `count` to 0.

jsuereth replied:
So, not actually true.   IF an error is thrown by one thread, what should happen in the other threads during usage?

We open-then-use.  Errors during the `use` thread should (possibly) prevent new threads from opening the same resource (instead they should wait for the resource to be closed).

Atry replied:
If an error is thrown by one thread, it should decrease the reference count by one, and check if closing underlying resource. Nothing else should happen.

Atry commented on src/main/scala/resource/SharedResource.scala:
```diff
@@ -0,0 +1,59 @@
+import java.util.concurrent.atomic.AtomicInteger
+
+final class SharedResource[A : Resource : Manifest](opener: => A) extends AbstractManagedResource[A] {
+	private val resource = implicitly[Resource[A]]
```
Could we change `val` to `def`, in order to avoid duplicated fields `resource` and `ev`

jsuereth replied:
sure

Atry commented on src/main/scala/resource/SharedResource.scala:
```diff
@@ -0,0 +1,59 @@
+		if (cnt == 0) {
+			// Mark the resource as closed, while we clean up behind the scenes.
+			// If the CAS fails, it means someone else has already opened the resource again.
+			ref.compareAndSet(Some(handle), None)
```
`compareAndSet` checks object reference equality, which is `eq` in Scala.
A newly created `Some` instance always does not `eq` to the one in `ref`.

We should store `null` or the `handle` itself, instead of an `Option`.

jsuereth replied:
Good point.  With the error-channel I need to have, this is becoming complicated.  I may store 'Object', actually, and it'll either be an error or a resource.....

jsuereth commented:
Ok @Atry I addressed all your comments.  This should highlight what I mean about the error channel/error handling.  I'd like to have a  sensible error strategy for sharing before releasing this code.

Atry commented:
I don't think we should close the underlying resource directly when the `for` block body throws an exception

jsuereth commented:
@Atry Why?   That's what we do in the non-shared case.    The issue with sharing a resource is you don't know if it's "contaiminated".    We could use the `Resource` type trait to determine if an exception reuqires the reosurce to close in a shared context, but I'd rather default to being 'careful' than possibly allowing lots of threads to use a "bad" resource.

In fact, it may actually be better to have some kind of "resource pool" implementation behind this where you draw a resource from the pool and the pool auto expands/collapses.  I feel like what's in there now is a bit naive/prone to failure in ways we don't really want to impose on the user.

Atry commented:
An exception thrown from a `for` block does not mean the resource is broken.

For example, suppose you have an image loaded from disk, and multiple threads render the same image to different render targets. One of these threads throws an exception does not break other theads.
